packages/flutter/lib/src/cupertino/refresh.dart

```diff
@@ -0,0 +1,514 @@
+
+  /// A builder that's called as this sliver's size changes and as the state
+  /// changes.
+  ///
+  /// A default simple Twitter-style pull-to-refresh indicator is provided if
+  /// not specified.
+  ///
+  /// Can be set to null, in which case nothing will be drawn in the overscrolled
+  /// space.
+  ///
+  /// Will not be called when the available space is 0 such as before any
+  /// overscroll.
+  final RefreshControlIndicatorBuilder builder;
+
+  /// Callback invoked when pulled by [refreshTriggerPullDistance].
+  ///
+  /// If provided, must return a [Future] which will keep the indicator in the
+  /// [RefreshIndicatorMode.refresh] state until the [Future] completes.
+  ///
+  /// Can be null in which case a single frame of [RefreshIndicatorMode.armed]
+  /// state will be drawn before going immediately to the [RefreshIndicatorMode.done]
+  /// where the sliver will start retracting.
+  final RefreshCallback onRefresh;
+
+  static const double _kDefaultRefreshTriggerPullDistance = 100.0;
+  static const double _kDefaultRefreshIndicatorExtent = 60.0;
+
+  /// Retrieve the current state of the CupertinoRefreshControl. The same as the
+  /// state that gets passed into the [builder] function. Used for testing.
+  @visibleForTesting
+  static RefreshIndicatorMode state(BuildContext context) {
+    final _CupertinoRefreshControlState state
+        = context.ancestorStateOfType(const TypeMatcher<_CupertinoRefreshControlState>());
+    return state.refreshState;
+  }
+
+  /// Builds a simple refresh indicator that fades in a bottom aligned down
+  /// arrow before the refresh is triggered, a [CupertinoActivityIndicator]
+  /// during the refresh and fades the [CupertinoActivityIndicator] away when
+  /// the refresh is done.
+  static Widget buildSimpleRefreshIndicator(BuildContext context,
+    RefreshIndicatorMode refreshState,
+    double pulledExtent,
+    double refreshTriggerPullDistance,
+    double refreshIndicatorExtent,
+  ) {
+    const Curve opacityCurve = const Interval(0.4, 0.8, curve: Curves.easeInOut);
+    return new Align(
+      alignment: Alignment.bottomCenter,
+      child: new Padding(
+        padding: const EdgeInsets.only(bottom: 16.0),
+        child: refreshState == RefreshIndicatorMode.drag
+            ? new Opacity(
+                opacity: opacityCurve.transform(
+                  min(pulledExtent / refreshTriggerPullDistance, 1.0)
+                ),
+                child: const Icon(
+                  CupertinoIcons.down_arrow,
+                  color: CupertinoColors.inactiveGray,
+                  size: 36.0,
+                ),
+              )
+            : new Opacity(
+                opacity: opacityCurve.transform(
+                  min(pulledExtent / refreshIndicatorExtent, 1.0)
+                ),
+                child: const CupertinoActivityIndicator(radius: 14.0),
+              ),
+      ),
+    );
+  }
+
+  @override
+  _CupertinoRefreshControlState createState() {
```

Comment:
this should eba one liner for consistency

Reply:
done